manager/controllers/scaling/metrics_cache.go

```diff
@@ -0,0 +1,217 @@
+package scaling
+
+import (
+	"container/list"
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	dfv1 "github.com/argoproj-labs/argo-dataflow/api/v1alpha1"
+	lru "github.com/hashicorp/golang-lru"
+	pmodel "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/expfmt"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	httpClient   *http.Client
+	metricsCache *lru.Cache
+)
+
+func init() {
+	t := http.DefaultTransport.(*http.Transport).Clone()
+	t.MaxIdleConns = 32
+	t.MaxConnsPerHost = 32
+	t.MaxIdleConnsPerHost = 32
+	t.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
+	httpClient = &http.Client{Timeout: 1 * time.Second, Transport: t}
+
+	metricsCache, _ = lru.New(1000)
+}
+
+type MetricsCacheHandler struct {
+	client client.Client
+
+	stepMap  map[string]*list.Element
```

Comment:
you could use `sync.Map` to make simpler

Reply:
I need to maintain a map and a queue and keep them consistent, which only need one lock. So it does not need a sync.Map.